ckanext/dataset_subscriptions/actions/whatsapp_notifications.py

```diff
@@ -96,3 +110,77 @@ def _validate_plugin_extras(extras):
+        if get_phonenumber(user):
+            return prepare_whatsapp_notifications(user)
+
+
+def prepare_whatsapp_notifications(user):
+    whatsapp_notifications_since = toolkit.config.get(
+            'ckanext.dataset_subscriptions.whatsapp_notifications_hours_since', 1)
+    whatsapp_notifications_since = int(whatsapp_notifications_since)
+    whatsapp_notifications_since = timedelta(hours=whatsapp_notifications_since)
+    whatsapp_notifications_since = (datetime.utcnow()
+                                    - whatsapp_notifications_since)
+    activity_stream_last_viewed = (
+            model.Dashboard.get(user['id']).activity_stream_last_viewed)
+    since = max(whatsapp_notifications_since, activity_stream_last_viewed)
+    return dms_whatsapp_notification_provider(user, since)
+
+
+def dms_whatsapp_notification_provider(user_dict, since):
+    context = {'model': model, 'session': model.Session,
+               'user': user_dict['id']}
+    activity_list = logic.get_action('dashboard_activity_list')(context, {})
+    dataset_activity_list = [activity for activity in activity_list
+                             if activity['user_id'] != user_dict['id']
+                             and 'package' in activity['activity_type']]
+    # We want a notification per changed dataset, not a list of all changes
+    timestamp_sorted_activity_list = sorted(dataset_activity_list,
+                                            key=lambda item: item['timestamp'])
+    deduplicated_activity_list = list({item["object_id"]:
+                                       item for item in timestamp_sorted_activity_list}.values())
+    activity_list_with_dataset_name = actions._add_dataset_name_to_activity_list(deduplicated_activity_list, context)
+    recent_activity_list = actions._filter_out_old_activites(activity_list_with_dataset_name, since)
+    if recent_activity_list:
+        user_phonenumber = get_phonenumber(user_dict)
+        return send_whatsapp_notification(recent_activity_list, user_phonenumber)
+
+
+def send_whatsapp_notification(activities, phonenumber):
+    from_nr = "whatsapp:" + SENDER_NR
+    to_nr = "whatsapp:" + phonenumber
+    header = toolkit.ungettext(
+        "{n} dataset you're following got recently updated in {site_title}",
+        "{n} datasets you're following got recently updated in {site_title}",
+        len(activities)).format(
+                site_title=toolkit.config.get('ckan.site_title'),
+                n=len(activities))
+    message_body = base.render(
+            'dataset-subscriptions_whatsapp_body.j2',
+            extra_vars={'activities': activities, 'header': header})
+    try:
+        message = client.messages.create(
+                                from_=from_nr,
+                                body=message_body,
+                                to=to_nr
+                                )
+    except TwilioRestException as e:
+        print('Whatsapp message send error: ' + e)
```

Comment:
use the logger to log exception details.
First in the top of the module to create logger for the module:
```
logger = logging.getLogger(__name__)
```
then:
```
logger.exception("Failed to send whatsapp message", exec_info=True)
```
exec_info will log the error message and ST.